rust/connlib/clients/shared/src/lib.rs

```diff
@@ -114,130 +98,39 @@ where
             }));
         }
 
-        Self::connect_inner(
-            &runtime,
-            tx,
-            api_url.try_into().map_err(|_| Error::UriError)?,
-            token,
-            device_id,
-            device_name_override,
-            os_version_override,
-            this.callbacks.clone(),
-            max_partition_time,
-        );
+        runtime.spawn({
+            let callbacks = callbacks.clone();
+
+            async move {
+                match connect(
+                    api_url,
+                    token,
+                    device_id,
+                    device_name_override,
+                    os_version_override,
+                    callbacks.clone(),
+                    max_partition_time,
+                )
+                .await
+                {
+                    Ok(never) => match never {},
+                    Err(e) => {
+                        tracing::error!("Tunnel failed: {e:#}");
+                        let _ = callbacks.on_disconnect(None);
+                    }
+                }
+            }
+        });
```

Comment:
This is still a bit ugly but I didn't want to make any further changes as part of this refactoring.